### PR TITLE
Run 90% of tests on compiled bundles (both development and production)

### PR DESCRIPTION
**Pull request by gaearon**
Adds **`yarn test-build`** and **`yarn test-build-prod`** that run on CI.
You can also run them locally after `yarn build <what you need> --type=NODE`.

Tests ending with `-test.internal.js` are blacklisted. They fail due to internal imports (e.g. `shared/ReactFeatureFlags`, mocks of local files, or direct file references). Some of them we can fix, but some will need to stay `.internal` forever which seems okay to me as long as it‘s not many.

Note this means that **if a non-internal test imports an internal file, it will not pass CI**. My suggestion for cases when you *need* an internal file (most commonly a feature flag or a mock) is to add a second `X-test.internal.js` file that only contains the "bundle-unfriendly" code. This way we can still run as many tests as we can (and even move tests between two files as some of them lose dependency on internals, e.g. if feature flag is removed).

This is all very new, and it's likely we'll uncover some pain points. I can't predict what those will be.
Let's try and see? We can always disable on CI if it's too painful.

### CODE REVIEW

**Comment by gaearon** [comment]
it's happening

About half the tests are enabled. Note that there's a *ton* of tests in DOMServerIntegration (which is currently disabled because it uses a feature flag). Once we fix that we should get a whole lot more passing.

Development bundles:

screen shot 2017-11-23 at 02 45 21

Production bundles:

screen shot 2017-11-23 at 02 42 23

CI passes too. 🙂

File: scripts/jest/config.source.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    '<rootDir>/scripts/rollup/shims/',
+    '<rootDir>/scripts/bench/',
+  ],
+  transform: {
+    '.*': require.resolve('./preprocessor.js'),
+  },
+  setupFiles: [require.resolve('./setupEnvironment.js')],
+  setupTestFrameworkScriptFile: require.resolve('./setupTests.js'),
+  testRegex: '/__tests__/.*(\\.js|\\.coffee|[^d]\\.ts)$',
+  moduleFileExtensions: ['js', 'json', 'node', 'coffee', 'ts'],
```
Comment (nhunzaker) [inline scripts/jest/config.source.js]: What is .node? I don't see it anywhere (other than .node.js)

Comment (gaearon) [reply scripts/jest/config.source.js]: Dunno. Used to be there. Maybe because it's one of Jest defaults? I can remove.

**Comment by gaearon** [comment]
>we could call it __unit__ and __acceptance__ or something.

Some counter arguments:

* "Private tests" should look "wrong". It should be obvious you shouldn't be adding more of them unless you really know what you're doing (e.g. adding a feature flag). Existing tests that are private (but don't have to be) should make you feel guilty. Legitimising them with a name like "__unit__" makes them seem like a good thing to write.
* The main distinction is whether they can run on bundles, not unit vs integration in traditional sense. When we fix the last dozen files, we won't have any unit tests at all. But even integration tests sometimes can't run on bundles (e.g. have to use an internal or mock something). Maybe there's some middle ground in naming though. 
* Related to the previous point, it should be easy to movie individual test cases between "public" and "private" versions of the same test file. Our goal is to maximize the number of test cases we run against bundles. So if a single case suddenly needs a feature flag it should be obvious where to put it (a "private" file next to its normal case). Separating them by directory will make it less obvious (and probably discourage people from moving individual tests back and forth).

**Comment by gaearon** [comment]
Maybe `.test.js` and `.test-internal.js`.

**Comment by trueadm** [comment]
@gaearon I think the naming is confusing me more. To me a private test seems like it's for testing internal private APIs/implementations. I think this comes from `public` vs `private` classes and methods etc. If this confused me, I'm sure it will confuse others too.

Okay, I'm going to suggest something radical but I think it actually makes *more* sense given what we're aiming to getting to: let me speak to you about first in a meeting.

**Comment by gaearon** [comment]
Updated to use `.internal.js` as convention for the time being.

**Comment by gaearon** [comment]
82% tests now running on bundles.

screen shot 2017-11-23 at 3 29 21 pm

**Comment by gaearon** [comment]
2312 of 2565: we have 90% tests running!

**Comment by gaearon** [comment]
This uncovered an issue with SSR warnings that print something different with flat bundles due to shared event system injection. I am changing the logic to be consistent in the next commit.

**Comment by Hoten** [comment]
I'm curious what sort of issues this testing uncovers. I hope @gaearon or others could update this thread with how this has helped!

**Comment by gaearon** [comment]
It enables things like https://github.com/facebook/react/pull/11967.

**Pull request merged**